# --------------------------------------- Workshop #2 -------------------------------------


* This workshop includes marked tasks that comprise 16% of your final mark in this module. 

* To complete the tasks, you need to apply data preprocessing methods discussed in Lecture 2. However, you may need to research to find solutions to some tasks. 

* You can duplicate the code and report cells if you need more than one code/report cell for your solution

# Tasks

## TASK 2.1: Download the adult dataset from Canvas, import it into Jupyter Notebook, and complete the following steps (Completing the report cell is required):

### a. Write a code to show how many Null values each column contains. Report the columns that contain Null values in the report cell (Hint: use the isna().sum() method to show the number of the Null values) (NOTE: Nan values are also considered as Null values) (1%). 

### b. Change the display setting and print the first 100 rows. Find columns that contain wrong data and report them in the report cell (1%). 


### c. We cannot inspect the entire data by eyes for columns containing wrong data. Explain in the report cell what method we could use to find all columns which contain wrong data (2%)


In [1]:
import pandas as pd;pd.set_option('display.max_columns', None)
import numpy as np
adult = pd.read_csv('adult.csv')
adult.head()


age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [2]:
##### TASK 2.1A #####
#find colunms with null values
adult.isna().sum()

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64

The number of identified null values in the adult dataset are identified as Workclass : 963, Occupation : 966 and native country : 274.


In [3]:
##### TASK 2.1b #####
# set display to show 100 rows
pd.set_option('display.max_rows', 100)

#print first 100 rows
print(adult.head(100))
#print('\n', adult.dtypes)

    age         workclass  fnlwgt     education  education-num  \
0    39         State-gov   77516     Bachelors             13   
1    50  Self-emp-not-inc   83311     Bachelors             13   
2    38           Private  215646       HS-grad              9   
3    53           Private  234721          11th              7   
4    28           Private  338409     Bachelors             13   
5    37           Private  284582       Masters             14   
6    49           Private  160187           9th              5   
7    52  Self-emp-not-inc  209642       HS-grad              9   
8    31           Private   45781       Masters             14   
9    42           Private  159449     Bachelors             13   
10   37           Private  280464  Some-college             10   
11   30         State-gov  141297     Bachelors             13   
12   23           Private  122272     Bachelors             13   
13   32           Private  205019    Assoc-acdm             12   
14   40   

The maximum number of rows displayed in the output is set to 100 and displayed

In [4]:
for i in adult.columns:
    print(adult[i].value_counts())
    print('\n', 50*'-')

age
36    1348
35    1337
33    1335
23    1329
31    1325
34    1303
28    1280
37    1280
30    1278
38    1264
32    1253
41    1235
27    1232
29    1223
24    1206
39    1206
25    1195
40    1187
22    1178
42    1165
26    1153
20    1113
43    1104
46    1097
21    1096
45    1096
47    1081
44    1067
19    1053
51     877
50     866
18     862
49     847
48     845
52     738
53     711
55     621
54     613
17     595
56     564
58     555
57     551
59     523
61     450
60     449
62     394
64     340
63     335
65     284
67     238
66     237
68     178
69     149
70     133
72     120
71     118
73     108
74      77
75      72
76      69
90      55
77      54
80      38
81      37
78      34
79      30
82      15
84      13
83      11
88       6
85       5
87       3
89       2
86       1
Name: count, dtype: int64

 --------------------------------------------------
workclass
Private             33906
Self-emp-not-inc     3862
Local-gov            3136
State-gov      

The code displays the occurrences of unique values in each column of a dataset such values may be considered outliers or incorrect, like the age 17 in the age column, and the occurrence of '?' values in the workclass, occupation, and native country column

##### TASK 2.1c #####

To identify columns that contain wrong data in the "adult" DataFrame we can use a combination of data cleaning techniques and statistical methods as listed below: 

(1) Use 'describe()' to detect outliers or anomalies in summary statistics.
(2) Use 'isnull()' to find columns with missing or null values.
(3) Apply 'value_counts()' to identify columns with low unique values or high missing percentages.
(4) Consider data profiling tools like pandas_profiling for detailed data quality reports.
(5) Employ data validation techniques for type, range, and custom rule validation.

Combining these methods helps pinpoint columns needing further investigation or cleaning due to anomalies, missing data, or inconsistent values.

## TASK 2.2: Complete the following tasks (Completing the report cell is required):

### a. Determine whether the columns which contain Null values are numerical, nominal, or ordinal variables and report it in the report cell (1%).

### b. Delete all rows which contain Null values (1%). Then print the number of Null values of all columns using the isna().sum() method. 

### c. Import the dataset again. Find the mode of the categorical (i.e. nominal or ordinal) columns which contain Null values and use it to fill their Null values (2%)



In [5]:
##### TASK 2.2a #####

# list of all columns containing and values
adult.isnull().sum()

age                 0
workclass         963
fnlwgt              0
education           0
education-num       0
marital-status      0
occupation        966
relationship        0
race                0
sex                 0
capital-gain        0
capital-loss        0
hours-per-week      0
native-country    274
income              0
dtype: int64

In [6]:
#extract only colunms with null variables
null_columns = adult.columns[adult.isna().any()].tolist()
print(null_columns)

['workclass', 'occupation', 'native-country']


In [7]:
#determine data type of colunms with null values
for column in null_columns:
    adult_type = adult[column].dtype
    print(f"{column}: {adult_type}")

workclass: object
occupation: object
native-country: object


The columns containing null values are nominal categorical variables. Workclass has 963,occupation has 966 values and native-country has 274 null values 


In [8]:
##### TASK 2.2b #####

# Delete all rows which contain Null values
adult_cleaned = adult.dropna()

# print the number of Null values of all columns using the isna().sum() method.
print(adult_cleaned.isna().sum())

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64


In [9]:
##### TASK 2.2c #####

import pandas as pd

# Import the dataset again.
adult2 = pd.read_csv('adult.csv')

# Find the mode of the categorical (i.e. nominal or ordinal) columns which contain Null values and use it to fill their Null values
categorical_columns = adult2.select_dtypes(include='object').columns
null_categorical_columns = categorical_columns[adult2[categorical_columns].isna().any()]

print(null_categorical_columns)
 

Index(['workclass', 'occupation', 'native-country'], dtype='object')


In [10]:
# Find the mode of each null categorical column and fill the Null values with the mode
for col in null_categorical_columns:
    mode = adult2[col].mode().iloc[0]
    adult2.loc[adult2[col].isna(), col] = mode
    
adult2.isna().sum()
    

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

## TASK 2.3. Select one of the columns which contains wrong data. Write a code to fill the wrong cells with an appropriate value. You have the freedom to determine what value you want to use to fill the wrong cells (2%). 

In [11]:
# Select one of the columns which contains wrong data.
selected_column = 'occupation'

# Write a code to fill the wrong cells with an appropriate value.
adult2[selected_column].replace('?', 'Unknown', inplace=True)
adult2[selected_column].unique()

array(['Adm-clerical', 'Exec-managerial', 'Handlers-cleaners',
       'Prof-specialty', 'Other-service', 'Sales', 'Craft-repair',
       'Transport-moving', 'Farming-fishing', 'Machine-op-inspct',
       'Tech-support', 'Unknown', 'Protective-serv', 'Armed-Forces',
       'Priv-house-serv'], dtype=object)

 The code replaces all occurrences of '?' in the 'occupation' column with the value 'Unknown'. It then prints the values in the 'occupation' column to confirm the replacement.

## TASK 2.4: Complete the following tasks: 

### a. Assume we want to predict the income column as the output. Use the correct method to encode this column (1%). 

### b. Select one ordinal column and encode it using the appropriate method (1%) . 

### c. Select one nominal column and encode it using the appropriate method (1%).

### d. After encoding the nominal column in the previous step, find a method to append the encoded data to the dataset (2%). 

### e. Normalise the numerical columns of the dataset (1%)

In [12]:
# 2.4a. We want to predict the income column as the output and Use the correct method to encode this column

import pandas as pd
from sklearn.preprocessing import LabelEncoder

#initialize the LabelEncoder
encoder = LabelEncoder()

# Select the column to be encoded.
selected_column = 'income'

# Encode the target 'income'
adult2['income_encoded'] = encoder.fit_transform(adult2['income'])
adult2.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  \
0          2174             0              40  United-States  <=50K   
1             0             0              13  United-States  <=50K   
2             0             0              40  United-States  <=50K   
3             0             0              40  United-States  <=50K   
4             0             0              40           Cuba  <=50K   

   income_encoded  
0               0  
1               0  
2               0  
3               0  
4               0

The output displays an additional new column 'income encoded' with values of the original income column encoded with assigned numerical values.

In [13]:
# 2.4b Select an ordinal column to encode

# Import the necessary libraries
import pandas as pd
import numpy as np
from sklearn.preprocessing import OrdinalEncoder

# initialize OrdinalEncoder
encoder = OrdinalEncoder()

# Fit the encoder to the selected column and transform the data
adult2['education'] = encoder.fit_transform(adult2['education'].values.reshape(-1,1))
adult2.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516        9.0             13   
1   50  Self-emp-not-inc   83311        9.0             13   
2   38           Private  215646       11.0              9   
3   53           Private  234721        1.0              7   
4   28           Private  338409        9.0             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  \
0          2174             0              40  United-States  <=50K   
1             0             0              13  United-States  <=50K   
2             0             0              40  United-States  <=50K   
3             0             0              40  United-States  <=50K   
4             0             0              40           Cuba  <=50K   

   income_encoded  
0               0  
1               0  
2               0  
3               0  
4               0

The 'education' column from the adult2 dataframe is selected and encoded using the OrdinalEncoder class from scikit-learn assigning numerical values based on the order of appearance in the unique values.

In [14]:
# 2.4c Select a nominal column to encode and append into dataset 


# Drop rows with missing values
adult_cleaned = adult.dropna()

# Encode nominal categorical variables using one-hot encoding
native_country_encoded = pd.get_dummies(adult_cleaned['native-country'], prefix='native-country')

# 2.4d Concatenate the encoded dataframe with the original dataframe
adult_encoded = pd.concat([adult_cleaned, native_country_encoded], axis=1)

# Drop the original 'native-country' column
adult_encoded.drop('native-country', axis=1, inplace=True)

# Print the first 5 rows of the encoded dataframe
print(adult_encoded.head())


   age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week income  native-country_?  \
0          2174             0              40  <=50K             False   
1             0             0     

The OneHotEncoder is used to encode the nominal column 'native-country' which splitted the values into new columns within the dataset. The pd.concat function was used to integrate the new nominal column into the dataset and the original column 'native-country' was dropped 

In [15]:
# Import StandardScaler
from sklearn.preprocessing import StandardScaler

# Define the scaler
scaler = StandardScaler()

#select numerical columns
numerical_columns = ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss']

# Fit the scaler to the numerical columns
adult[numerical_columns] = scaler.fit_transform(adult[numerical_columns])

# Display normalized dataframe
print(adult.head())

        age         workclass    fnlwgt  education  education-num  \
0  0.025996         State-gov -1.061979  Bachelors       1.136512   
1  0.828308  Self-emp-not-inc -1.007104  Bachelors       1.136512   
2 -0.046942           Private  0.246034    HS-grad      -0.419335   
3  1.047121           Private  0.426663       11th      -1.197259   
4 -0.776316           Private  1.408530  Bachelors       1.136512   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0      0.146932     -0.217127              40  United-States  <=50K  


The numerical columns 'age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', and 'hours-per-week' are normalized using StandardScaler thus maintaining a uniform scale across all features.

Word count : 351 words